### PR TITLE
Allow user to specify max iterations and make partial spectra fitting more flexible.

### LINKED ISSUE #22860: Improve FitPeaks' flexibility on partial spectra fitting
### Expected behavior

FitPeaks shall be flexible to fit partial spectra with peak center workspace and peak window workspace.

An option shall be added for user to determine the iteration of fitting.

### Actual behavior

FitPeaks refused to fit partial spectra with presence of peak center workspace and peak window workspace.

### PR DESCRIPTION
**Pull request by wdzhou**
**Description of work.**

There are two major improvements for FitPeaks.

1.  An optional output TableWorkspace is added for showing fit function parameters' uncertainties.
2. Both output TableWorksapces, for parameters and uncertainties, will have same number of rows as the number the spectra to fit.

**To test:**

1. Check new unit tests
2. Run FitPeaks on a multiple spectra workspace to fit peaks in a part of spectra.  Check both of the output TableWorkspace for fitted parameter value and uncertainties.

Fixes #22860.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/FitPeaks.cpp [hunk 1]
```diff
@@ -75,6 +78,24 @@ size_t PeakFitResult::getNumberPeaks() const {
   return m_function_parameters_vector.size();
 }
 
+//----------------------------------------------------------------------------------------------
+/** get the fitting error of a particular parameter
+ * @brief PeakFitResult::getParameterError
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/FitPeaks.cpp]: remove empty docs

File: Framework/Algorithms/src/FitPeaks.cpp [hunk 2]
```diff
@@ -75,6 +78,24 @@ size_t PeakFitResult::getNumberPeaks() const {
...
+ * @param iparam
+ * @return
+ */
+double PeakFitResult::getParameterError(size_t ipeak, size_t iparam) const {
+  return m_function_errors_vector[ipeak][iparam];
+}
+
+//----------------------------------------------------------------------------------------------
+/** get the fitted value of a particular parameter
+ * @brief PeakFitResult::getParameterValue
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/FitPeaks.cpp]: remove empty docs

File: Framework/Algorithms/src/FitPeaks.cpp [hunk 3]
```diff
@@ -410,6 +463,19 @@ std::map<std::string, std::string> FitPeaks::validateInputs() {
...
 
+  // check inputs for uncertainty (fitting error)
+  std::string error_table_name =
+      getPropertyValue("OutputParameterFitErrorsWorkspace");
+  if (error_table_name != "") {
+    bool use_raw_params = getProperty("RawPeakParameters");
+    if (!use_raw_params) {
+      std::string msg =
+          "FitPeaks must output RAW peak parameters if fitting error "
+          "is chosen to be output.";
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/FitPeaks.cpp]: remove `.` at the end of the sentence

File: Framework/Algorithms/src/FitPeaks.cpp [hunk 4]
```diff
@@ -1573,7 +1705,7 @@ double FitPeaks::fitFunctionSD(IAlgorithm_sptr fit,
   fit->setProperty("Function", fitfunc);
   fit->setProperty("InputWorkspace", dataws);
   fit->setProperty("WorkspaceIndex", static_cast<int>(wsindex));
-  fit->setProperty("MaxIterations", 50); // magic number
+  fit->setProperty("MaxIterations", m_fitIterations); // magic number
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/FitPeaks.cpp]: I like that this is being exposed as a parameter to the user

File: Framework/Algorithms/src/FitPeaks.cpp [hunk 5]
```diff
@@ -2019,8 +2191,12 @@ void FitPeaks::getRangeData(size_t iws,
 }
 
 //----------------------------------------------------------------------------------------------
-// find 2 local minima: draw a line as background to reduce
-// find 1 local minima: a flat background
+/** Reduce Y value with given background function
+ * @brief FitPeaks::reduceByBackground
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/FitPeaks.cpp]: remove empty parameter docs

File: Framework/Algorithms/test/FitPeaksTest.h [hunk 6]
```diff
@@ -691,7 +864,90 @@ class FitPeaksTest : public CxxTest::TestSuite {
...
+
+    AnalysisDataService::Instance().addOrReplace(workspace_name, center_ws);
+
+    return workspace_name;
+  }
+
+  //--------------------------------------------------------------------------------------------------------------
+  /** create a fit window workspace compatibel to the workspace created by
+   * createTestData()
+   * @brief genFitWindowWorkspace
```
Comment (peterfpeterson) [inline Framework/Algorithms/test/FitPeaksTest.h]: remove empty docstring

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Lots of minor changes. One suggestion (for a separate PR) is to change the tests to use [CreateSampleWorkspace](http://docs.mantidproject.org/nightly/algorithms/CreateSampleWorkspace-v1.html) with `Function='Powder Diffraction'` which could simplify the tests.

File: Framework/Algorithms/src/FitPeaks.cpp [hunk 7]
```diff
@@ -1614,10 +1745,9 @@ double FitPeaks::fitFunctionSD(IAlgorithm_sptr fit,
 
   // Retrieve result
   std::string fitStatus = fit->getProperty("OutputStatus");
-  double chi2 = DBL_MAX;
+  double chi2(DBL_MAX);
```
[the remark this answers is not in the document]
Comment (wdzhou) [reply Framework/Algorithms/src/FitPeaks.cpp]: double chi2(DBL_MAX) is C++ standard; 
double chi2{DB_MAX} is introduced in C++ 11.

**Review by peterfpeterson — DISMISSED** [review]
Looks good now

File: Framework/Algorithms/src/FitPeaks.cpp [hunk 8]
```diff
@@ -410,6 +460,19 @@ std::map<std::string, std::string> FitPeaks::validateInputs() {
     }
   }
 
+  // check inputs for uncertainty (fitting error)
+  std::string error_table_name =
```
Comment (soininen) [inline Framework/Algorithms/src/FitPeaks.cpp]:
```suggestion
  const std::string error_table_name =
```

File: Framework/Algorithms/src/FitPeaks.cpp [hunk 9]
```diff
@@ -410,6 +460,19 @@ std::map<std::string, std::string> FitPeaks::validateInputs() {
...
+      getPropertyValue("OutputParameterFitErrorsWorkspace");
+  if (error_table_name != "") {
+    bool use_raw_params = getProperty("RawPeakParameters");
```
Comment (soininen) [inline Framework/Algorithms/src/FitPeaks.cpp]:
```suggestion
    const bool use_raw_params = getProperty("RawPeakParameters");
```

File: Framework/Algorithms/src/FitPeaks.cpp [hunk 10]
```diff
@@ -410,6 +460,19 @@ std::map<std::string, std::string> FitPeaks::validateInputs() {
     }
   }
 
+  // check inputs for uncertainty (fitting error)
+  std::string error_table_name =
+      getPropertyValue("OutputParameterFitErrorsWorkspace");
+  if (error_table_name != "") {
```
Comment (soininen) [inline Framework/Algorithms/src/FitPeaks.cpp]:
```suggestion
  if (!error_table_name.empty()) {
```

File: Framework/Algorithms/src/FitPeaks.cpp [hunk 11]
```diff
@@ -410,6 +460,19 @@ std::map<std::string, std::string> FitPeaks::validateInputs() {
...
+    bool use_raw_params = getProperty("RawPeakParameters");
+    if (!use_raw_params) {
+      std::string msg =
```
Comment (soininen) [inline Framework/Algorithms/src/FitPeaks.cpp]:
```suggestion
      `const` std::string msg =
```

File: Framework/Algorithms/src/FitPeaks.cpp [hunk 12]
```diff
@@ -706,16 +773,34 @@ void FitPeaks::processInputPeakCenters() {
...
       m_partialSpectra = true;
-    else
-      throw std::invalid_argument(
-          "Input peak center workspace has wrong number of spectra.");
+    } else {
+      // a case indicating programming error
+      g_log.error() << "Peak center workspace has "
+                    << peak_center_ws_spectra_number << " spectra; "
+                    << "Input workspace has "
+                    << m_inputMatrixWS->getNumberHistograms() << " spectra;"
```
Comment (soininen) [inline Framework/Algorithms/src/FitPeaks.cpp]:
```suggestion
                    << m_inputMatrixWS->getNumberHistograms() << " spectra; "
```

File: Framework/Algorithms/src/FitPeaks.cpp [hunk 13]
```diff
@@ -1614,10 +1740,9 @@ double FitPeaks::fitFunctionSD(IAlgorithm_sptr fit,
 
   // Retrieve result
   std::string fitStatus = fit->getProperty("OutputStatus");
-  double chi2 = DBL_MAX;
+  double chi2{DBL_MAX};
```
Comment (soininen) [inline Framework/Algorithms/src/FitPeaks.cpp]:
```suggestion
  double chi2{std::numeric_limits<double>::max()};
```

File: Framework/Algorithms/src/FitPeaks.cpp [hunk 14]
```diff
@@ -1786,56 +1911,93 @@ void FitPeaks::generateOutputPeakPositionWS() {
...
+  std::vector<std::string> param_vec;
+  if (m_rawPeaksTable) {
+    std::vector<std::string> peak_params = m_peakFunction->getParameterNames();
+    for (size_t i = 0; i < peak_params.size(); ++i)
+      param_vec.push_back(peak_params[i]);
+  } else {
+    param_vec.push_back("centre");
+    param_vec.push_back("width");
+    param_vec.push_back("height");
+    param_vec.push_back("intensity");
```
Comment (soininen) [inline Framework/Algorithms/src/FitPeaks.cpp]: All `push_back`s could be replaced by `emplace_back`

File: Framework/Algorithms/src/FitPeaks.cpp [hunk 15]
```diff
@@ -1786,56 +1911,93 @@ void FitPeaks::generateOutputPeakPositionWS() {
...
+  }
+  // background
+  for (size_t iparam = 0; iparam < m_bkgdFunction->nParams(); ++iparam)
+    param_vec.push_back(m_bkgdFunction->parameterName(iparam));
```
Comment (soininen) [inline Framework/Algorithms/src/FitPeaks.cpp]:
```suggestion
    param_vec.emplace_back(m_bkgdFunction->parameterName(iparam));
```

File: Framework/Algorithms/src/FitPeaks.cpp [hunk 16]
```diff
@@ -1786,56 +1911,93 @@ void FitPeaks::generateOutputPeakPositionWS() {
...
+
+  // parameter value table
+  m_fittedParamTable =
+      WorkspaceFactory::Instance().createTable("TableWorkspace");
+  setupParameterTableWorkspace(m_fittedParamTable, param_vec, true);
+
+  // for error workspace
+  std::string fiterror_table_name =
```
Comment (soininen) [inline Framework/Algorithms/src/FitPeaks.cpp]:
```suggestion
  const std::string fiterror_table_name =
```

File: Framework/Algorithms/src/FitPeaks.cpp [hunk 17]
```diff
@@ -1786,56 +1911,93 @@ void FitPeaks::generateOutputPeakPositionWS() {
...
+      getPropertyValue("OutputParameterFitErrorsWorkspace");
+  // do nothing if user does not specifiy
+  if (fiterror_table_name == "") {
```
Comment (soininen) [inline Framework/Algorithms/src/FitPeaks.cpp]:
```suggestion
  if (fiterror_table_name.empty()) {
```

File: Framework/Algorithms/test/FitPeaksTest.h [hunk 18]
```diff
@@ -54,6 +54,87 @@ class FitPeaksTest : public CxxTest::TestSuite {
...
+  //----------------------------------------------------------------------------------------------
+  /** test fit a single peak in partial spectra from a multiple spectra
+   * workspace
+   * the peak positions are given by the peak position workspace and thus the
+   * peak fit windows
+   * @brief test_singlePeaksPartialSpectra
+   */
+  void test_singlePeaksPartialSpectra() {
+    // Generate input workspace
+    std::string data_ws_name("Test1Data");
```
Comment (soininen) [inline Framework/Algorithms/test/FitPeaksTest.h]: `const`?

File: Framework/Algorithms/test/FitPeaksTest.h [hunk 19]
```diff
@@ -54,6 +54,87 @@ class FitPeaksTest : public CxxTest::TestSuite {
...
+    std::vector<double> peakparvalues;
+    gen_PeakParameters(peakparnames, peakparvalues);
+
+    // create a 1-value peak index vector for peak (0) at X=5
+    std::vector<int> peak_index_vec;
+    peak_index_vec.push_back(0);
+    std::string ws_name("peakcenter1");
+    std::string peak_center_ws_name =
+        genPeakCenterWorkspace(peak_index_vec, ws_name);
+    std::string fit_window_ws_name =
```
Comment (soininen) [inline Framework/Algorithms/test/FitPeaksTest.h]: The above `string`s could be declared `const`.

File: Framework/Algorithms/test/FitPeaksTest.h [hunk 20]
```diff
@@ -691,7 +864,91 @@ class FitPeaksTest : public CxxTest::TestSuite {
...
+              << " bins"
+              << "\n";
+    for (size_t i = 0; i < center_ws->getNumberHistograms(); ++i) {
+      for (size_t j = 0; j < peak_index_vec.size(); ++j) {
+        int peak_index = peak_index_vec[j];
+        double peak_center(-1);
+        if (peak_index == 0)
+          peak_center = 5.0;
+        else
+          peak_center = 10.0;
```
Comment (soininen) [inline Framework/Algorithms/test/FitPeaksTest.h]: This if-else could be replaced by
```c++
const double peak_centre = peak_index == 0 ? 5.0 : 10.0;
```

File: Framework/Algorithms/test/FitPeaksTest.h [hunk 21]
```diff
@@ -691,7 +864,91 @@ class FitPeaksTest : public CxxTest::TestSuite {
...
+        boost::dynamic_pointer_cast<MatrixWorkspace>(
+            WorkspaceCreationHelper::create2DWorkspaceWithValuesAndXerror(
+                nhist, nbins, ishist, xval, yval, eval, dxval, maskedws));
+
+    std::cout << "Center workspace has " << center_ws->readX(0).size()
+              << " bins"
+              << "\n";
+    for (size_t i = 0; i < center_ws->getNumberHistograms(); ++i) {
+      for (size_t j = 0; j < peak_index_vec.size(); ++j) {
+        int peak_index = peak_index_vec[j];
```
Comment (soininen) [inline Framework/Algorithms/test/FitPeaksTest.h]:
```suggestion
        const int peak_index = peak_index_vec[j];
```

File: Framework/Algorithms/test/FitPeaksTest.h [hunk 22]
```diff
@@ -691,7 +864,91 @@ class FitPeaksTest : public CxxTest::TestSuite {
...
+   * createTestData()
+   * @brief genFitWindowWorkspace :: generate a matrix work for peak fitting
+   * window
+   * @param peak_index_vec :: vector for peak indexes
+   * @param workspace_name :: name of the output workspace registered to ADS
+   */
+  std::string genFitWindowWorkspace(std::vector<int> &peak_index_vec,
+                                    const std::string &workspace_name) {
+    // create the empty workspace containing 3 spectrum
+    size_t num_peaks = peak_index_vec.size();
```
Comment (soininen) [inline Framework/Algorithms/test/FitPeaksTest.h]:
```suggestion
    const size_t num_peaks = peak_index_vec.size();
```

File: Framework/Algorithms/test/FitPeaksTest.h [hunk 23]
```diff
@@ -691,7 +864,91 @@ class FitPeaksTest : public CxxTest::TestSuite {
...
+            WorkspaceCreationHelper::create2DWorkspace(
+                3, static_cast<int>(num_peaks) * 2));
+    for (size_t i = 0; i < center_ws->getNumberHistograms(); ++i) {
+      for (size_t j = 0; j < peak_index_vec.size(); ++j) {
+        int peak_index = peak_index_vec[j];
+        double peak_center(-1);
+        if (peak_index == 0)
+          peak_center = 5.0;
+        else
+          peak_center = 10.0;
```
Comment (soininen) [inline Framework/Algorithms/test/FitPeaksTest.h]: See the above comments about the if-else and `const`.

**Review by soininen — CHANGES_REQUESTED** [review]

**Review by soininen — APPROVED** [review]

**Comment by martyngigg** [comment]
There seem to be some compiler warnings on clang: http://builds.mantidproject.org/job/pull_requests-osx/25459/warnings14Result/

**Pull request merged**